src/pow.rs

```diff
@@ -270,3 +346,207 @@ fn count_leading_zeros(data: &[u8]) -> u8 {
 
     leading_zeros
 }
+
+/// The faucet will dynamically adjust the difficulty based on the
+/// current balance of the faucet to make it increasingly difficult
+/// to retrieve funds from the faucet. The actual equation for this
+/// is:
+///
+/// ```math
+/// f_unclamped(x)=(M-m)(1-\log_{q}\frac{x}{b})+m
+/// f_clamped(x) = \max(m,\min(M,f_unclamped(x)))
+/// f(x) = f_clamped(x)
+/// ```
+///
+/// where:
+///
+/// - `x` is the current balance in BTC
+/// - `M` is the maximum difficulty
+/// - `m` is the minimum difficulty
+/// - `b` is the minimum balance in BTC
+/// - `q` is the amount emitted per request in BTC
+///
+/// # Guarantees
+///
+/// This function guarantees that the difficulty will be between `min_difficulty`
+/// and `max_difficulty` given that the correctness assumptions are met.
+///
+/// # Correctness
+///
+/// For this function correctly output, you must ensure:
+///
+/// - `per_emission` > `Amount::ONE_BTC`, ideally >2 BTC due to the way the curve
+///   functions
+/// - `min_difficulty` <= `max_difficulty`
+/// - `max_difficulty`, `min_difficulty`, `min_balance`, `q` and `x` all > 0
+///
+/// # Expected clamping points
+///
+/// f(x) = M at x = b
+/// f(x) = m at x = bq.
+pub fn calculate_difficulty(x: f32, big_m: f32, m: f32, b: f32, q: f32) -> f32 {
+    // optimisation for when the balance is less than or equal to the min balance
+    if x <= b {
+        return big_m;
+    }
+
+    let fval = (big_m - m) * (1.0 - (x / b).log(q)) + m;
```

Comment:
This will panic if `b == 0`. Recommend checked division here.

Reply:
Actually, the division seems to gracefully handle zero divisor. `(x/b)` is `inf` and `fval` evaluates to `-inf` which when clamped, correctly returns minimum difficulty `m`.

Added a test case for when `b == 0`.